### ISSUE TITLE
cache mutation detector causes memory/cpu pressure at the end of long e2e runs (like pull-kubernetes-e2e-gce-etcd3)

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#47135]
seeing a spike in ns cleanup failure: https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=Couldn%27t%20delete%20ns#5c121dd9347b4a1cbd0d

looks like it started the morning of 6/6

currently failing 5% of gce-etcd3 jobs

cc @kubernetes/sig-release-misc @kubernetes/sig-testing-misc

**Comment by liggitt** [comment]
@smarterclayton @derekwaynecarr fyi

**Comment by liggitt** [comment]
@mikedanese

**Comment by liggitt** [comment]
possible namespace deletion change ~2 days ago: https://github.com/kubernetes/kubernetes/pull/46765

**Comment by smarterclayton** [comment]
It's *possible* it's the initializers as well, I'm delivering a second fix for that area in https://github.com/kubernetes/kubernetes/pull/47138 in case it's correlated

**Comment by derekwaynecarr** [comment]
@liggitt -- the linked pr seems like it would have not changed the e2e
behavior unless that e2e env did not previously have tpr enabled.  it does
look like it removed static caching of discovery though, but we need to do
that anyway now i guess...

**Comment by liggitt** [comment]
There was a typo in the enablement check that meant dynamic discovery was never run  previously

**Comment by foxish** [comment]
/cc. This is stopping every other batched merge.

**Comment by liggitt** [comment]
Agree, seems to be blocking most PRs I look at

**Comment by liggitt** [comment]
I spent a while looking at the PRs around the time of the spike and nothing jumped out at me. I'm really not sure what sig to point this at

**Comment by liggitt** [comment]
cc @kubernetes/sig-api-machinery-bugs @kubernetes/sig-node-bugs as the two sigs most involved in namespace and graceful deletion

**Comment by yujuhong** [comment]
In https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/batch/pull-kubernetes-e2e-gce-etcd3/34991, test timed out around 14:24 because namespace could not be deleted (`e2e-tests-clientset-hrxpx `)
```
I0607 14:24:30.078] [1mSTEP[0m: Destroying namespace "e2e-tests-clientset-hrxpx" for this suite.
I0607 14:24:30.078] Jun  7 14:24:29.543: INFO: namespace: e2e-tests-clientset-hrxpx, resource: bindings, ignored listing per whitelist
I0607 14:24:30.079] Jun  7 14:24:30.073: INFO: Couldn't delete ns: "e2e-tests-clientset-hrxpx": namespace e2e-tests-clientset-hrxpx was not deleted with limit: timed out waiting for the condition, namespace is empty but is not yet removed (&errors.errorString{s:"namespace e2e-tests-clientset-hrxpx was not deleted with limit: timed out waiting for the condition, namespace is empty but is not yet removed"})
```
From the apiserver.log, kubelet deleted the only pod `podfda938ed-4bc6-11e7-98d7-0242ac110002` in the namespace around 21:21:27, well before the timeout.
```
I0607 21:19:29.167012       5 wrap.go:42] DELETE /api/v1/namespaces/e2e-tests-clientset-hrxpx: (46.038259ms) 200 [[e2e.test/v1.8.0 (linux/amd64) kubernetes/6301725] 104.198.229.226:48228]
I0607 21:21:25.167376       5 wrap.go:42] DELETE /api/v1/namespaces/e2e-tests-clientset-hrxpx/pods: (61.400101ms) 200 [[kube-controller-manager/v1.8.0 (linux/amd64) kubernetes/6301725/system:serviceaccount:kube-system:namespace-controller] [::1]:60878]
...
I0607 21:21:26.454099       5 wrap.go:42] DELETE /apis/autoscaling/v1/namespaces/e2e-tests-clientset-hrxpx/horizontalpodautoscalers: (424.983302ms) 200 [[kube-controller-manager/v1.8.0 (linux/amd64) kubernetes/6301725/system:serviceaccount:kube-system:namespace-controller] [::1]:60878]
I0607 21:21:27.083397       5 wrap.go:42] DELETE /apis/rbac.authorization.k8s.io/v1alpha1/namespaces/e2e-tests-clientset-hrxpx/rolebindings: (59.669782ms) 200 [[kube-controller-manager/v1.8.0 (linux/amd64) kubernetes/6301725/system:serviceaccount:kube-system:namespace-controller] [::1]:60878]
I0607 21:21:27.316572       5 wrap.go:42] DELETE /api/v1/namespaces/e2e-tests-clientset-hrxpx/pods/podfda938ed-4bc6-11e7-98d7-0242ac110002: (48.864897ms) 200 [[kubelet/v1.8.0 (linux/amd64) kubernetes/6301725] 35.192.22.196:35513]
I0607 21:21:27.332655       5 wrap.go:42] GET /api/v1/namespaces/e2e-tests-clientset-hrxpx/pods/podfda938ed-4bc6-11e7-98d7-0242ac110002: (8.507988ms) 404 [[kubelet/v1.8.0 (linux/amd64) kubernetes/6301725] 35.192.22.196:35513]
```

The controller manager also acknowledged the pod deletion in its log, but I have no idea why it said `unexpected items still remain in namespace: e2e-tests-clientset-hrxpx for gvr: { v1 pods}`, and whether that's something of interest. /cc @nikhiljindal who last touched this code path.
```
I0607 21:21:35.522238       5 taint_controller.go:341] Noticed pod deletion: types.NamespacedName{Namespace:"e2e-tests-clientset-hrxpx", Name:"podfda938ed-4bc6-11e7-98d7-0242ac110002"}
I0607 21:21:40.226220       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-clientset-hrxpx" (30.5502549s)
E0607 21:21:40.226281       5 namespace_controller.go:148] unexpected items still remain in namespace: e2e-tests-clientset-hrxpx for gvr: { v1 pods}
I0607 21:21:51.269383       5 serviceaccounts_controller.go:184] Finished syncing namespace "e2e-tests-clientset-hrxpx" (1.668µs)
I0607 21:21:51.818769       5 tokens_controller.go:249] syncServiceAccount(e2e-tests-clientset-hrxpx/default), service account deleted, removing tokens
I0607 21:24:00.418124       5 namespaced_resources_deleter.go:496] namespace controller - deleteAllContent - namespace: e2e-tests-clientset-hrxpx
I0607 21:24:29.536174       5 namespaced_resources_deleter.go:522] namespace controller - deleteAllContent - namespace: e2e-tests-clientset-hrxpx, estimate: 0
I0607 21:24:30.702227       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-clientset-hrxpx" (31.446592756s)
I0607 21:24:30.738082       5 serviceaccounts_controller.go:184] Finished syncing namespace "e2e-tests-clientset-hrxpx" (19.782µs)
I0607 21:27:05.484057       5 namespace_controller.go:171] Namespace has been deleted e2e-tests-clientset-hrxpx
I0607 21:27:05.484071       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-clientset-hrxpx" (30.992µs)
```

**Comment by liggitt** [comment]
seeing super slow namespace deletion, and the kubelet trying to mount volumes for a pod 2 minutes after it has observed the pod being deleted:

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46796/pull-kubernetes-e2e-gce-etcd3/35035/
```
e2e test deletes the pod:
I0608 02:53:29.432385       5 wrap.go:42] DELETE /api/v1/namespaces/e2e-tests-configmap-nscgm/pods/pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002: (395.468602ms) 200 [[e2e.test/v1.8.0 (linux/amd64) kubernetes/a56f5b2] 104.197.171.12:41426]

kubelet observes the deletion:
I0608 02:53:29.148132    2580 kubelet.go:1873] SyncLoop (DELETE, "api"): "pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002_e2e-tests-configmap-nscgm(a4ddbccc-4bf5-11e7-a943-42010a800002)"
I0608 02:53:29.369733    2580 kubelet.go:1867] SyncLoop (REMOVE, "api"): "pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002_e2e-tests-configmap-nscgm(a4ddbccc-4bf5-11e7-a943-42010a800002)"                                                                              │
I0608 02:53:29.369805    2580 kubelet.go:2051] Failed to delete pod "pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002_e2e-tests-configmap-nscgm(a4ddbccc-4bf5-11e7-a943-42010a800002)", err: pod not found        

kubelet tries to update pod status and fails:
W0608 02:53:29.616132    2580 status_manager.go:448] Failed to update status for pod "pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002_e2e-tests-configmap-nscgm(a4ddbccc-4bf5-11e7-a943-42010a800002)": Operation cannot be fulfilled on pods "pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002": StorageError: invalid object, Code: 4, Key: /registry/pods/e2e-tests-configmap-nscgm/pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: a4ddbccc-4bf5-11e7-a943-42010a800002, UID in object meta: 

namespace delete is initiated by the e2e test:
I0608 02:53:30.736101       5 wrap.go:42] DELETE /api/v1/namespaces/e2e-tests-configmap-nscgm: (849.011439ms) 200 [[e2e.test/v1.8.0 (linux/amd64) kubernetes/a56f5b2] 104.197.171.12:41426] 

namespace controller is busy (slowly) syncing other namespaces:
I0608 02:53:52.796968       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-statefulset-5w7r3" (1m8.738234563s)
I0608 02:53:58.886617       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-multi-az-f397n" (1m2.69736313s)  
I0608 02:54:02.287418       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-svc-latency-qt2h7" (5m2.323954438s)
I0608 02:54:06.238942       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-init-container-hk0gg" (1m5.214912452s)
I0608 02:54:13.957240       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-job-dbtvt" (59.588699428s) 
I0608 02:54:52.189797       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-initializers-zsx5g" (53.303140508s)
I0608 02:55:01.578033       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-clientset-hj173" (55.33900797s) 
I0608 02:55:02.189367       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-network-vjznp" (59.901765047s)
I0608 02:55:11.229484       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-projected-9s9xv" (57.272057097s)
...

the kubelet is still trying to mount volumes for the pod it observed be deleted:
E0608 02:55:31.641116    2580 kubelet.go:1602] Unable to mount volumes for pod "pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002_e2e-tests-configmap-nscgm(a4ddbccc-4bf5-11e7-a943-42010a800002)": timeout expired waiting for volumes to attach/mount for pod "e2e-tests-configmap-nscgm"/"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002". list of unattached/unmounted volumes=[configmap-volume default-token-62kcn]; skipping pod
E0608 02:55:31.641171    2580 pod_workers.go:182] Error syncing pod a4ddbccc-4bf5-11e7-a943-42010a800002 ("pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002_e2e-tests-configmap-nscgm(a4ddbccc-4bf5-11e7-a943-42010a800002)"), skipping: timeout expired waiting for volumes to attach/mount for pod "e2e-tests-configmap-nscgm"/"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002". list of unattached/unmounted volumes=[configmap-volume default-token-62kcn]

kubelet cannot record events because the namespace is terminating:
I0608 02:55:31.641512    2580 server.go:351] Event(v1.ObjectReference{Kind:"Pod", Namespace:"e2e-tests-configmap-nscgm", Name:"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002", UID:"a4ddbccc-4bf5-11e7-a943-42010a800002", APIVersion:"v1", ResourceVersion:"13346", FieldPath:""}): type: 'Warning' reason: 'FailedSync' Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "e2e-tests-configmap-nscgm"/"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002". list of unattached/unmounted volumes=[configmap-volume default-token-62kcn]
I0608 02:55:31.641536    2580 server.go:351] Event(v1.ObjectReference{Kind:"Pod", Namespace:"e2e-tests-configmap-nscgm", Name:"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002", UID:"a4ddbccc-4bf5-11e7-a943-42010a800002", APIVersion:"v1", ResourceVersion:"13346", FieldPath:""}): type: 'Warning' reason: 'FailedMount' Unable to mount volumes for pod "pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002_e2e-tests-configmap-nscgm(a4ddbccc-4bf5-11e7-a943-42010a800002)": timeout expired waiting for volumes to attach/mount for pod "e2e-tests-configmap-nscgm"/"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002". list of unattached/unmounted volumes=[configmap-volume default-token-62kcn]
I0608 02:55:32.108092       5 wrap.go:42] POST /api/v1/namespaces/e2e-tests-configmap-nscgm/events: (543.724µs) 403 [[kubelet/v1.8.0 (linux/amd64) kubernetes/a56f5b2] 35.192.22.196:54345]
E0608 02:55:32.203810    2580 event.go:200] Server rejected event '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002.14c60601f61e6d74", GenerateName:"", Namespace:"e2e-tests-configmap-nscgm", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, InvolvedObject:v1.ObjectReference{Kind:"Pod", Namespace:"e2e-tests-configmap-nscgm", Name:"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002", UID:"a4ddbccc-4bf5-11e7-a943-42010a800002", APIVersion:"v1", ResourceVersion:"13346", FieldPath:""}, Reason:"FailedSync", Message:"Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod \"e2e-tests-configmap-nscgm\"/\"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002\". list of unattached/unmounted volumes=[configmap-volume default-token-62kcn]", Source:v1.EventSource{Component:"kubelet", Host:"e2e-35035-minion-group-tkqf"}, FirstTimestamp:v1.Time{Time:time.Time{sec:63632487331, nsec:641191796, loc:(*time.Location)(0x57fe020)}}, LastTimestamp:v1.Time{Time:time.Time{sec:63632487331, nsec:641191796, loc:(*time.Location)(0x57fe020)}}, Count:1, Type:"Warning"}': 'events "pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002.14c60601f61e6d74" is forbidden: unable to create new content in namespace e2e-tests-configmap-nscgm because it is being terminated.' (will not retry!)
I0608 02:55:32.336387       5 wrap.go:42] POST /api/v1/namespaces/e2e-tests-configmap-nscgm/events: (6.384372ms) 403 [[kubelet/v1.8.0 (linux/amd64) kubernetes/a56f5b2] 35.192.22.196:54345]
E0608 02:55:32.352676    2580 event.go:200] Server rejected event '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002.14c60601f61ce5d4", GenerateName:"", Namespace:"e2e-tests-configmap-nscgm", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, InvolvedObject:v1.ObjectReference{Kind:"Pod", Namespace:"e2e-tests-configmap-nscgm", Name:"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002", UID:"a4ddbccc-4bf5-11e7-a943-42010a800002", APIVersion:"v1", ResourceVersion:"13346", FieldPath:""}, Reason:"FailedMount", Message:"Unable to mount volumes for pod \"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002_e2e-tests-configmap-nscgm(a4ddbccc-4bf5-11e7-a943-42010a800002)\": timeout expired waiting for volumes to attach/mount for pod \"e2e-tests-configmap-nscgm\"/\"pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002\". list of unattached/unmounted volumes=[configmap-volume default-token-62kcn]", Source:v1.EventSource{Component:"kubelet", Host:"e2e-35035-minion-group-tkqf"}, FirstTimestamp:v1.Time{Time:time.Time{sec:63632487331, nsec:641091540, loc:(*time.Location)(0x57fe020)}}, LastTimestamp:v1.Time{Time:time.Time{sec:63632487331, nsec:641091540, loc:(*time.Location)(0x57fe020)}}, Count:1, Type:"Warning"}': 'events "pod-configmaps-a4d08477-4bf5-11e7-be23-0242ac110002.14c60601f61ce5d4" is forbidden: unable to create new content in namespace e2e-tests-configmap-nscgm because it is being terminated.' (will not retry!)

the e2e test times out after 5 minutes waiting for the namespace to delete and starts scanning for remaining content to report on:
I0608 02:58:31.226107       5 wrap.go:42] GET /apis/rbac.authorization.k8s.io/v1alpha1/namespaces/e2e-tests-configmap-nscgm/roles: (7.108828ms) 200 [[e2e.test/v1.8.0 (linux/amd64) kubernetes/a56f5b2] 104.197.171.12:41426]                                               │
I0608 02:58:31.233988       5 wrap.go:42] GET /apis/extensions/v1beta1/namespaces/e2e-tests-configmap-nscgm/daemonsets: (6.01749ms) 200 [[e2e.test/v1.8.0 (linux/amd64) kubernetes/a56f5b2] 104.197.171.12:41426]                                                           │
I0608 02:58:31.271034       5 wrap.go:42] GET /apis/extensions/v1beta1/namespaces/e2e-tests-configmap-nscgm/replicationcontrollers: (725.384µs) 404 [[e2e.test/v1.8.0 (linux/amd64) kubernetes/a56f5b2] 104.197.171.12:41426]                                              r│
I0608 02:58:31.343285       5 wrap.go:42] GET /api/v1/namespaces/e2e-tests-configmap-nscgm/persistentvolumeclaims: (28.94551ms) 200 [[e2e.test/v1.8.0 (linux/amd64) kubernetes/a56f5b2] 104.197.171.12:41426]                                                               │

namespace controller finally starts to clean up:
I0608 02:59:05.237681       5 namespaced_resources_deleter.go:496] namespace controller - deleteAllContent - namespace: e2e-tests-configmap-nscgm
I0608 02:59:47.452937       5 namespaced_resources_deleter.go:522] namespace controller - deleteAllContent - namespace: e2e-tests-configmap-nscgm, estimate: 0
I0608 02:59:48.997657       5 wrap.go:42] DELETE /api/v1/namespaces/e2e-tests-configmap-nscgm: (1.13331104s) 200 [[kube-controller-manager/v1.8.0 (linux/amd64) kubernetes/a56f5b2/system:serviceaccount:kube-system:namespace-controller] [::1]:48750]                     │
I0608 02:59:49.252454       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-configmap-nscgm" (44.516925479s) 
```

**Comment by mikedanese** [comment]
Ref https://github.com/kubernetes/kubernetes/pull/45666

**Comment by mikedanese** [comment]
cc @kubernetes/sig-release-bugs @dchen1107 this has been blocking the SQ for 3 days and we still haven't pinned the root cause.

**Comment by foxish** [comment]
The batched merges haven't been blocked by this for the past 12 hours or so. The only PR that may have fixed this that I see is https://github.com/kubernetes/kubernetes/pull/47086

**Comment by dchen1107** [comment]
Yes, this is our top 1 issue. So far, ~50% of batch merges failed due to this issue.

**Comment by foxish** [comment]
Worth looking at https://prow.k8s.io/?type=presubmit&job=pull-kubernetes-e2e-gce-etcd3

**Comment by liggitt** [comment]
Disabling the gzip filter appears to have fixed the namespace deletion problem, though I'm not sure why. The merge of that feature corresponds with when the flakes began.

Revert is in https://github.com/kubernetes/kubernetes/pull/47191

**Comment by spiffxp** [comment]
/remove-sig testing

keeping any eye on this, but this seems like an issue with app-specific code, not test-infra

**Comment by derekwaynecarr** [comment]
a few ideas on things we could do:

the node controller no longer force deletes pods when nodes are unhealthy (https://github.com/kubernetes/kubernetes/pull/36017).

if we see message: "namespace %v was not deleted with limit: %v, pods remaining: %v" it could mean that the node in this case never sent the final delete call to API server.  It is also possible that the node is unhealthy, and the node controller would have set the reason on the pod to "NodeLost".  In fact, right now, its impossible to delete a namespace if any of the nodes for pods in that namespace unresponsive.  you are required to manually force delete the pod...  anyway, if we can dump the yaml of the pod when stuck terminating, it would help probably let us know if this is the source...

**Comment by yujuhong** [comment]
@derekwaynecarr in the one failure I looked at (https://github.com/kubernetes/kubernetes/issues/47135#issuecomment-306961946), kubelet did delete the pod (and got 404 on the next GET). Is it possible to surface more information in these cases as well?

**Comment by liggitt** [comment]
I am still observing namespace deletion taking longer and longer throughout the e2e run, and am seeing some failures (though they have improved with the revert of the gzip filter). I am speculating that the apiserver is CPU-bound and is very close to the timeout point for ns cleanup, which means any increase is likely to cause flakes/failures.

@wojtek-t @kubernetes/sig-scalability-bugs have we seen any spikes in apiserver CPU usage the past few days (since 6/6 especially)?

**Comment by wojtek-t** [comment]
> @wojtek-t @kubernetes/sig-scalability-bugs have we seen any spikes in apiserver CPU usage the past few days (since 6/6 especially)?

I don't see anything suspicious in the cpu usage in scalability tests. Though we are still not monitoring it in regular e2e tests. @ncdc was trying to do that but it was reverted unfortunately and I think we didn't have time to get back to it.

**Comment by liggitt** [comment]
@smarterclayton mentioned adding an e2e that captures a CPU profile... that might be helpful

**Comment by smarterclayton** [comment]
Basically we should be able to hit the master and hit /debug/pprof/profile
and stream the output to disk, then run go tool pprof on it and get the
top20.  Will try this afternoon if i get past other bugs

**Comment by dchen1107** [comment]
After reverting  "add gzip compression to GET and LIST requests" #47191, the batch merge rate acceptable. Now we have another issue #47350 which causes namespace deletion is constantly failing. That is more severe one. But assigning @yujuhong to root cause this.

**Comment by dchen1107** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/47362

**Comment by mikedanese** [comment]
Ok looking into this some more I pulled together a timeline for one ns delete failure:
```
23:26:06 e2e test deletes ns
23:28:25 ns controller starts first sync of ns
23:28:46 kubelets delete 2 pods (apiserver sees this)
23:28:56 ns controller finish first sync (FAILS as expected due to 2 pods being in the ns)
23:30:55 ns controller starts second sync loop
23:31:06 e2e test times out ns delete
23:31:37 ns controller finishes second sync loop with SUCCESS
```
Observations:
1. Namespace deletes take 30+ seconds and always fail on the first try if pods are in the namespace.
2. A fractional change in apiserver latency causes that 30 seconds to blowup because a single sync loop loops over all GVKs and lists.
3. The maximum number of concurrent loops is 5.
4. simple math => max namespace syncs per minute is 5
5. we run >300 e2e tests in `pull-kubernetes-e2e-gce-etcd3`

Proposal:
bump the number of concurrent namespace syncs and optimize the sync loop after 1.7.

**Comment by yujuhong** [comment]
What I observed is mostly consistent with @mikedanese's comment above.

Here is a timeline from one of the failures:
```
- 19:38:31.821937 apiserver records a DELETE request for the namespace.
- 19:41:18.651159 apiserver records a DELETE request for the pod in the namespace.
- 19:41:18.651159 apiserver records a DELETE request from the kubelet to delete the pod.
- 19:43:33.625159 namespace is finally deleted, but the test timed out just seconds ago.
```
Snippet from the ns controller:
```
I0612 19:39:51.678904       5 namespaced_resources_deleter.go:499] namespace controller - deleteAllContent - namespace: e2e-tests-svc-latency-sndfc
I0612 19:41:27.635712       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-svc-latency-sndfc" (1m36.011848831s)
E0612 19:41:27.635779       5 namespace_controller.go:148] unexpected items still remain in namespace: e2e-tests-svc-latency-sndfc for gvr: { v1 pods}
I0612 19:43:11.024090       5 namespaced_resources_deleter.go:499] namespace controller - deleteAllContent - namespace: e2e-tests-svc-latency-sndfc
I0612 19:43:33.393260       5 namespaced_resources_deleter.go:525] namespace controller - deleteAllContent - namespace: e2e-tests-svc-latency-sndfc, estimate: 0
I0612 19:43:33.739718       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-svc-latency-sndfc" (23.097121394s)
I0612 19:45:19.377325       5 namespace_controller.go:171] Namespace has been deleted e2e-tests-svc-latency-sndfc
I0612 19:45:19.377332       5 namespace_controller.go:166] Finished syncing namespace "e2e-tests-svc-latency-sndfc" (23.306µs)
 ```

1. syncing the namesapce took a long time. The pod deletion request was issued almost 3 minutes after the original namespace deletion request.
2. The first sync failed due to [`unexpected items still remain in namespace: e2e-tests-svc-latency-sndfc for gvr: { v1 pods}`](https://github.com/kubernetes/kubernetes/blob/v1.8.0-alpha.0/pkg/controller/namespace/deletion/namespaced_resources_deleter.go#L484). This was because the controller used [TerminationGracePeriod as the estimate time](https://github.com/kubernetes/kubernetes/blob/v1.8.0-alpha.0/pkg/controller/namespace/deletion/namespaced_resources_deleter.go#L549) (which was zero for the test) and the pod didn't have any finalizer.
3. Due to the error and lack of estimate information in (2), the namespace was requeued for the next sync. I *think* this may add extra delay to the wait time. 
4. By the time the namespace was picked up and finished syncing again, the test had already timed out.

Like @mikedanese suggested, the sync loop couldn't handle the volume of namespaces in the suite. Another issue may be that whether we should fix the behavior in (2) and (3) since we'd need to allow time for kubelet to delete the pod anyway regardless of the existence of finalizers. /cc @derekwaynecarr

**Comment by mikedanese** [comment]
Cluster is showing a slight uptick in flakes after merge of fix:

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=Couldn%27t%20delete%20ns#5c121dd9347b4a1cbd0d

Potentially 10 is not enough?

**Comment by gmarek** [comment]
It looks like namespace controller needs some attention. If anyone is willing to go and try to profile/optimize it it'd be great. Otherwise I'll do it next week.

**Comment by liggitt** [comment]
it's certainly possible there's an issue with the controller, but I was seeing increasing latency on API requests throughout the e2e run... it's possible the API server just gets slower.

**Comment by liggitt** [comment]
time throughout the run is graphed at https://github.com/kubernetes/kubernetes/pull/46796#issuecomment-307981276

**Comment by derekwaynecarr** [comment]
@liggitt @yujuhong - per https://github.com/kubernetes/kubernetes/issues/47135#issuecomment-307937470 , i am wondering if `estimateGracefulTerminationForPods` should never return 0, and instead have a minEstimate of ~2s?

**Comment by yujuhong** [comment]
> @liggitt @yujuhong - per #47135 (comment) , i am wondering if estimateGracefulTerminationForPods should never return 0, and instead have a minEstimate of ~2s?

I think it makes sense to have some default non-zero estimate. kubelet also sets a minimum grace period of 2s itself, so expecting immediate termination is not practical. 
This may not help with the namespace deletion slowness (which looks more like the api request latency issue), but should at least avoid unnecessary error messages.

**Comment by dchen1107** [comment]
Re: https://github.com/kubernetes/kubernetes/issues/47135#issuecomment-308135523

@liggitt Do we have more stats on increasing latency on API requests? @gmarek @wojtek-t Given https://github.com/kubernetes/kubernetes/issues/44003 was fixed for cluster perf dashboard, can we have AB comparison on API latency between 1.6 and 1.7? cc/ @k8s-mirror-api-machinery-misc 

If there is no latency downgrade from above measurements, I suggested we close this issue for 1.7. There is no need to artificially change the real production code to fix the test infrastructure given it is not that mature. 

On another hand, we should open a new one to figure out the bottleneck on the namespace slowness, and generate the benchmark for this release, so that we can easily decide if there is a performance regression for the future releases.

**Comment by liggitt** [comment]
> @liggitt Do we have more stats on increasing latency on API requests?

Discovery API requests are taking ~5-10ms at the beginning of the e2e run and ~200-500ms at the end. I'm trying to capture memory and CPU profiles throughout the e2e run in https://github.com/kubernetes/kubernetes/pull/47527 to help determine the root cause

**Comment by dchen1107** [comment]
@liggitt I updated at https://github.com/kubernetes/kubernetes/issues/47446#issuecomment-308584924

**Comment by crimsonfaith91** [comment]
One of my PRs are affected too: https://github.com/kubernetes/kubernetes/pull/47469

Part of the build log:
```
I0614 16:28:49.906] • Failure in Spec Teardown (AfterEach) [611.751 seconds]
I0614 16:28:49.906] [k8s.io] InitContainer
I0614 16:28:49.906] /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:619
I0614 16:28:49.906]   should not start app containers and fail the pod if init containers fail on a RestartNever pod [AfterEach]
I0614 16:28:49.907]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/common/init_container.go:452
I0614 16:28:49.907] 
I0614 16:28:49.907]   Jun 14 16:28:49.898: Couldn't delete ns: "e2e-tests-init-container-91xzd": namespace e2e-tests-init-container-91xzd was not deleted with limit: timed out waiting for the condition, pods remaining: 2, pods missing deletion timestamp: 2 (&errors.errorString{s:"namespace e2e-tests-init-container-91xzd was not deleted with limit: timed out waiting for the condition, pods remaining: 2, pods missing deletion timestamp: 2"})
```

I will follow closely to this issue, and retest the PR frequently.

**Comment by dchen1107** [comment]
Please my latest update at https://github.com/kubernetes/kubernetes/issues/47446#issuecomment-309189490

I am suggesting to close this one too, but I will leave @liggitt to make the final call. Thanks!

**Comment by liggitt** [comment]
since the test infra changes have resolved the blocking issue, and the cause of the performance hit during this test is known, I'm fine with lowering severity and moving out of the milestone. I would like to retitle this and keep it open to briefly investigate if we can make the mutation detector more of a constant cost operation throughout the test run

**Comment by dchen1107** [comment]
Thanks, @liggitt
